### PR TITLE
Add vectorized buffer initialization for groupby and join

### PR DESCRIPTION
**Pull request by ienkovich**

### CODE REVIEW

File: QueryEngine/JoinHashTable/Runtime/HashJoinRuntime.cpp [hunk 1]
```diff
@@ -86,6 +87,67 @@ inline int64_t translate_str_id_to_outer_dict(const int64_t elem,
...
+
+/**
+ * For AVX512 target we perform manual vectorization to use non-temporal stores.
+ */
+__attribute__((target("avx512f"), optimize("no-tree-vectorize"))) void
+init_hash_join_buff_cpu(int32_t* groups_buffer,
+                        const int32_t invalid_slot_val,
+                        const int64_t start,
+                        const int64_t end) {
+  int64_t pos = start;
```
Comment (fexolm) [inline QueryEngine/JoinHashTable/Runtime/HashJoinRuntime.cpp:111]: we could probably want to move this implementation to smth like fast_memset function to be able to use it elsewhere

File: QueryEngine/QueryMemoryInitializer.cpp [hunk 2]
```diff
@@ -149,6 +151,80 @@ inline std::vector<std::vector<int64_t>> get_col_frag_offsets(
...
+    return 0;
+  } else {
+    assert(!(v & 1));
+    return const_log2(v / 2) + 1;
+  }
+}
+
+// Return a min number of rows required to fill a pattern
+// divisible by vec_size.
+template <size_t vec_size>
```
Comment (fexolm) [inline QueryEngine/QueryMemoryInitializer.cpp:165]: does this really optimizes smth? I mean, template argument + constexpr

File: QueryEngine/QueryMemoryInitializer.cpp [hunk 3]
```diff
@@ -149,6 +151,80 @@ inline std::vector<std::vector<int64_t>> get_col_frag_offsets(
...
+         (reinterpret_cast<uint64_t>(sample_ptr) & 0x3F));
+  // Align buffers.
+  int64_t align_bytes = ((64ULL - reinterpret_cast<uint64_t>(dst)) & 0x3F);
+  memcpy(dst, sample_ptr, align_bytes);
+
+  int8_t* align_dst = dst + align_bytes;
+  const int8_t* align_sample = sample_ptr + align_bytes;
+  size_t rem = dst_size - align_bytes;
+  size_t rem_scalar = rem % sample_size;
+  size_t rem_vector = rem - rem_scalar;
```
Comment (fexolm) [inline QueryEngine/QueryMemoryInitializer.cpp:197]: could we also move it to fast_memcpy and just use the same implementation, like in default version, but with fast_memcpy and 512bit pattern?

**Review by fexolm — APPROVED** [review]
LGTM, just wondering, if we could expose that optimization to use it elsewhere.

**Comment by dchigarev** [comment]
@ienkovich Hash table fill for baseline join also suffers from non-sequential initialization on CPU, do we need a new PR to fix this, or you'll do it here?
https://github.com/intel-ai/omniscidb/blob/46c011dc68632a975df55f02649fbac1f6021350/QueryEngine/JoinHashTable/Runtime/HashJoinRuntime.cpp#L363-L368

File: QueryEngine/QueryMemoryInitializer.cpp (hunk 2 above)
Comment (ienkovich) [reply QueryEngine/QueryMemoryInitializer.cpp:165]: This moves computations into compile time. Don't think it affects performance significantly.

**Comment by ienkovich** [comment]
> @ienkovich Hash table fill for baseline join also suffers from non-sequential initialization on CPU, do we need a new PR to fix this, or you'll do it here?

This will have to be handled separately.

File: QueryEngine/QueryMemoryInitializer.cpp (hunk 3 above)
Comment (ienkovich) [reply QueryEngine/QueryMemoryInitializer.cpp:197]: I don't fully understand your suggestion here.

File: QueryEngine/JoinHashTable/Runtime/HashJoinRuntime.cpp (hunk 1 above)
Comment (ienkovich) [reply QueryEngine/JoinHashTable/Runtime/HashJoinRuntime.cpp:111]: Didn't see any places for re-use yet.

**Pull request merged**